File: go/internal/csvdiff/sections.go
package csvdiff

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
)

// assemble turns a completed join into the capped report sections, and writes
// the uncapped CSV exports when --export-dir is set.
func assemble(meta EngineMeta, joined Joined, dupA, dupB Section, opt Options, compared []string) (EngineResult, error) {
	cols := append(append([]string{}, opt.Key...), compared...)

	if opt.ExportDir != "" {
		if err := export(joined, opt, compared, cols); err != nil {
			return EngineResult{}, err
		}
	}

	return EngineResult{
		Meta:    meta,
		Counts:  joined.Counts,
		Columns: joined.Columns,
		Changed: CapSection(opt.Key, joined.Changed, opt.MaxRows),
		Added:   CapSection(cols, joined.Added, opt.MaxRows),
		Removed: CapSection(cols, joined.Removed, opt.MaxRows),
		DupA:    dupA,
		DupB:    dupB,
	}, nil
}

func export(joined Joined, opt Options, compared, cols []string) error {
	dir := opt.ExportDir
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("cannot create export-dir %s: %w", dir, err)
	}
	if err := writeCSV(filepath.Join(dir, "added.csv"), cols, joined.Added); err != nil {
		return err
	}
	if err := writeCSV(filepath.Join(dir, "removed.csv"), cols, joined.Removed); err != nil {
		return err
	}

	both := append([]string{}, opt.Key...)
	for _, c := range compared {
		both = append(both, c+" (A)", c+" (B)")
	}
	keySize := len(opt.Key)
	rows := make([][]any, 0, len(joined.ChangedA))
	for i := range joined.ChangedA {
		ar, br := joined.ChangedA[i], joined.ChangedB[i]
		row := make([]any, 0, len(both))
		for j := range keySize {
			row = append(row, ar[j])
		}
		for j := range compared {
			row = append(row, ar[keySize+j], br[keySize+j])
		}
		rows = append(rows, row)
	}
	return writeCSV(filepath.Join(dir, "changed.csv"), both, rows)
}

func writeCSV(path string, header []string, rows [][]any) error {
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("cannot write %s: %w", path, err)
	}
	defer f.Close()

	bw := bufio.NewWriterSize(f, 1<<20)
	w := csv.NewWriter(bw)
	if err := w.Write(header); err != nil {
		return err
	}
	rec := make([]string, len(header))
	for _, row := range rows {
		for i := range rec {
			rec[i] = ""
			if i < len(row) {
				if v, ok := row[i].(*string); ok && v != nil {
					rec[i] = *v
				}
			}
		}
		if err := w.Write(rec); err != nil {
			return err
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return err
	}
	return bw.Flush()
}

File: rust/src/sections.rs
//! Capping the report sections and writing the uncapped CSV exports.

use std::fs;
use std::path::Path;

use crate::contract::{Cell, EngineMeta, EngineResult, Section, Val};
use crate::error::{Error, Result};
use crate::options::Options;
use crate::rowstore::Joined;

/// Turns a completed join into the capped report sections, and writes the
/// uncapped CSV exports when `--export-dir` is set.
pub fn assemble(
    meta: EngineMeta,
    joined: Joined,
    dup_a: Section,
    dup_b: Section,
    opt: &Options,
    compared: &[String],
) -> Result<EngineResult> {
    let mut cols = opt.key.clone();
    cols.extend_from_slice(compared);

    if let Some(dir) = &opt.export_dir {
        export(Path::new(dir), &joined, opt, compared, &cols)?;
    }

    Ok(EngineResult {
        meta,
        counts: joined.counts,
        columns: joined.columns,
        changed: Section::capped(opt.key.clone(), joined.changed, opt.max_rows),
        added: Section::capped(cols.clone(), joined.added, opt.max_rows),
        removed: Section::capped(cols, joined.removed, opt.max_rows),
        dup_a,
        dup_b,
    })
}

fn export(
    dir: &Path,
    joined: &Joined,
    opt: &Options,
    compared: &[String],
    cols: &[String],
) -> Result<()> {
    fs::create_dir_all(dir)
        .map_err(|e| Error::new(format!("cannot create export-dir {}: {e}", dir.display())))?;
    write_cells(&dir.join("added.csv"), cols, &joined.added)?;
    write_cells(&dir.join("removed.csv"), cols, &joined.removed)?;

    // The changed export puts the two sides side by side, which the sparse
    // report payload cannot show.
    let mut both = opt.key.clone();
    for c in compared {
        both.push(format!("{c} (A)"));
        both.push(format!("{c} (B)"));
    }
    let key_size = opt.key.len();
    let rows: Vec<Vec<Val>> = joined
        .changed_a
        .iter()
        .zip(&joined.changed_b)
        .map(|(ar, br)| {
            let mut row: Vec<Val> = ar[..key_size].to_vec();
            for i in 0..compared.len() {
                row.push(ar[key_size + i].clone());
                row.push(br[key_size + i].clone());
            }
            row
        })
        .collect();
    write_rows(&dir.join("changed.csv"), &both, &rows)
}

fn write_cells(path: &Path, header: &[String], rows: &[Vec<Cell>]) -> Result<()> {
    let values: Vec<Vec<Val>> = rows
        .iter()
        .map(|row| row.iter().map(|c| c.value().map(str::to_string)).collect())
        .collect();
    write_rows(path, header, &values)
}

/// Writes an uncapped export file. Shared with the Polars engine, which builds
/// its rows straight from the frame rather than from a [`Joined`].
pub fn write_rows(path: &Path, header: &[String], rows: &[Vec<Val>]) -> Result<()> {
    let mut w = csv::Writer::from_path(path)
        .map_err(|e| Error::new(format!("cannot write {}: {e}", path.display())))?;
    w.write_record(header)?;
    let mut record: Vec<&str> = Vec::with_capacity(header.len());
    for row in rows {
        record.clear();
        for i in 0..header.len() {
            record.push(row.get(i).and_then(Option::as_deref).unwrap_or(""));
        }
        w.write_record(&record)?;
    }
    w.flush()?;
    Ok(())
}
